[dagster-postgres] Fix QueuePool overflow

axelwas opened the pull request:
## Summary & Motivation

Aims to fix issue #26400 where the QueuePool used by PostgresEventLogStorage overflows when manually launching a materialisation of a big number of assets.

The root cause seems to be the graphql client in the webserver launching multiple queries to fetch/update status of the many assets. It causes a spike in connections which overflows the pool.

Both `PostgresRunStorage` and `PostgresScheduleStorage` share the same default pool config. I chose not to change anything as we didn't experience any issues with those.

## How I Tested These Changes

Vendored dagster-postgres into our Dagster project. Launching manual materializations of 2000+ assets works, despite being quite slow.

## Changelog

Removed the cap on PostgresEventLogStorage QueuePool by setting max_overflow to -1 (see SQLAlchemy [docs](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.max_overflow)).

axelwas commented:
I haven't tried setting a large default pool but monitored the max number of connections in the pool when materializing 2000+ assets. Number of connections didn't exceed 15 so just setting `max_overflow` to 20 should work.

Went for the quick fix as those settings were already hardcoded for all (event log, run storage and schedule storage). 
I didn't find a reason to make it configurable but I believe it could be achived by specifying `max_overflow` in dagster.yaml and passing `inst_data` where `PostgresEventLogStorage` is [instanciated](https://github.com/dagster-io/dagster/blob/27bbe4faa79b70e5065e503ed15f00ef2710a1a5/python_modules/libraries/dagster-postgres/dagster_postgres/storage.py#L49).

Do you want me to go ahead and do that ?

axelwas commented:
Another option would be to add a `db-pool-max-overflow` to `dagster-webserver` CLI command [here](https://github.com/dagster-io/dagster/blob/acc5d1a671170b949ba9657ae0665a8b17e3becd/python_modules/dagster-webserver/dagster_webserver/cli.py#L184) and propagate the change to all `optimize_for_webserver` methods in mysql/postgres manager classes. This is how it was done for the `pool_recycle` param.

Keen to get your thoughts on this @sbquinlan 😄

sbquinlan commented:
Yes, I think adding a set of optional pool params to the PostgresConfig https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/storage/config.py#L35 seems like the right approach here and then threading it through.

axelwas commented:
As the initial issue was only occuring in the webserver component and affects any sqlalchemy pool (mysql or postgres), I choose to add a param to the CLI command `dagster-webserver`. This implementation fits better with the current codebase as this is how the `pool_recycle` param is passed.

I modified the helm chart and added a test to test the new Helm value.

All tests are passing locally.

sbquinlan commented:
Thanks for doing this again. I have some test failures for you to look at:

```[2025-02-15T01:15:42Z] FAILED dagster_postgres_tests/test_instance.py::test_statement_timeouts - TypeError: DagsterInstance.optimize_for_webserver() missing 1 required positional argument: 'max_overflow'
[2025-02-15T01:15:42Z] FAILED dagster_postgres_tests/test_instance.py::test_configured_other_schema - TypeError: DagsterInstance.optimize_for_webserver() missing 1 required positional argument: 'max_overflow'
``

sbquinlan commented:
Also:

```
[2025-02-15T01:22:07Z] FAILED dagster_mysql_tests/test_wait_timeout.py::test_pool_recycle_greater_than_wait_timeout - TypeError: MySQLRunStorage.optimize_for_we
bserver() missing 1 required positional argument: 'max_overflow'
[2025-02-15T01:22:07Z] FAILED dagster_mysql_tests/test_wait_timeout.py::test_pool_recycle_less_than_wait_timeout - TypeError: MySQLRunStorage.optimize_for_webse
rver() missing 1 required positional argument: 'max_overflow'
```

Seems to be just test code that needs updating.

sbquinlan commented:
last one is pyright that found basically the same things https://buildkite.com/dagster/unit-tests/builds/1458#01950732-c874-48d2-8f02-b19286b20aaf

sbquinlan reviewed (changes requested):
just requesting for inbox zero on my side.

axelwas commented:
Sorry for the back and forth.

Updated the test code, looks good now.

sbquinlan reviewed (approved):
Nice thanks!

sbquinlan commented:
This will make it out next week by the way, we already cut the branches for this week's release.

axelwas commented:
Fantastic thanks for the review @sbquinlan
